lib/block_view.go

```diff
@@ -2987,24 +3067,74 @@ func _isEntryImmatureBlockReward(utxoEntry *UtxoEntry, blockHeight uint32, param
 	return false
 }
 
-func _verifySignature(txn *MsgBitCloutTxn) error {
-	// Compute a hash of the transaction
+func (bav *UtxoView) _verifySignature(txn *MsgBitCloutTxn, blockHeight uint32) error {
+	// Compute a hash of the transaction.
 	txBytes, err := txn.ToBytes(true /*preSignature*/)
 	if err != nil {
 		return errors.Wrapf(err, "_verifySignature: Problem serializing txn without signature: ")
 	}
 	txHash := Sha256DoubleHash(txBytes)
-	// Convert the txn public key into a *btcec.PublicKey
-	txnPk, err := btcec.ParsePubKey(txn.PublicKey, btcec.S256())
+
+	// We make a local copy of the signature as it should be read-only.
+	sigCopy := append([]byte{}, txn.Signature...)
+	// Verify that the control byte is within the allowed range.
+	// DERControlByte means txn was signed with the owner's private key.
+	// Otherwise, txn was signed with a derived key and we will attempt
+	// to recover the public key from the signature.
+	sigByte := sigCopy[0]
+	if !(sigByte >= DERControlByte && sigByte <= SIGMaxByte) {
+		return fmt.Errorf("_verifySignature: Signature control byte outside of allowed range")
+	}
+
+	// Reset the control byte to the DER control byte. Verify signature encoding.
+	sigCopy[0] = DERControlByte
+	sig, err := btcec.ParseDERSignature(sigCopy, btcec.S256())
 	if err != nil {
-		return errors.Wrapf(err, "_verifySignature: Problem parsing public key: ")
+		return errors.Wrapf(err, "_verifySignature: Problem parsing signature: ")
 	}
-	// Verify that the transaction is signed by the specified key.
-	if txn.Signature == nil || !txn.Signature.Verify(txHash[:], txnPk) {
-		return RuleErrorInvalidTransactionSignature
+
+	// We check for public key solution iteration for this transaction. This
+	// allows us to recover public key from the signature.
+	// If the transaction was signed with the owner key, iteration is equal to 0.
+	// If the transaction was signed with a derived key, iteration is between 1-4.
+	iteration := sigByte - DERControlByte
+	if iteration == 0 {
```

Comment:
I think this logic would be a lot nicer if it was more like:

```
if isDerivedKey {
// something
} else {
// something
}
```

This sigByte stuff is too clever and makes the code a lot harder to read. I'd prefer to take an efficiency hit on an extraData bool rather than embed this in the signature itself.

Reply:
Please let me know if you still think we'd want the additional field in extraData after my comment on the public key iteration.